contrib/firebaseio/src/main/java/io/FirebaseSource.java

```diff
@@ -0,0 +1,397 @@
+   * @return A FirebaseSource that uses only a {@link ValueEventListener}
+   */
+  public static <K> FirebaseSource<K> valueListenerOf(
+      Class<K> clazz,
+      FirebaseAuthenticator auther,
+      Query query){
+    return new FirebaseSource<>(clazz, auther, query, false, true);
+  }
+
+  /**
+   * @param clazz determines the type of data the {@link DataSnapshot}s received from Firebase
+   * servers will be unmarshalled into, and serialized and deserialized. Any POJO which is
+   * serializable as a JSON Object by an {@link ObjectWriter} will work.
+   * @param auther Used to authenticate to Firebase
+   * @param query The {@link Query} which will create events passed on to the pipeline.
+   * @return A FirebaseSource that uses both a {@link ChildEventListener} and a
+   * {@link ValueEventListener}
+   */
+  public static <K> FirebaseSource<K> of(
+      Class<K> clazz,
+      FirebaseAuthenticator auther,
+      Query query){
+    return new FirebaseSource<>(clazz, auther, query, true, true);
+  }
+
+  /**
+   * @param clazz Class Reference for the type to deserialize {@link DataSnapshot}s into.
+   * @param auther Authenticator
+   * @param query Any {@link Query} reference.
+   * @param listenForChildEvents Adds a {@link ChildEventListener} during runtime
+   * @param listenForValueEvents Adds a {@link ValueEventListener} during runtime
+   */
+  protected FirebaseSource(
+      Class<T> clazz,
+      FirebaseAuthenticator auther,
+      Query query,
+      boolean listenForChildEvents,
+      boolean listenForValueEvents){
+    if (query.toString().equals(query.getRef().getRoot().toString())){
+      throw new IllegalArgumentException("Cannot use a reference to the root of a Firebase "
+          + "repository. Since FirebaseSource uses a ref at the root of the repository to store"
+          + "timestamps, this will result in an endless event loop, which will livelock your"
+          + "pipeline.");
+    }
+    this.clazz = clazz;
+    this.auther = auther;
+    this.queryString = query.getRef().toString();
+    this.listenForChildEvents = listenForChildEvents;
+    this.listenForValueEvents = listenForValueEvents;
+  }
+
+  protected Firebase getRef(){
+    return new Firebase(this.queryString);
+  }
+
+  @SuppressWarnings("unchecked")
+  @Override
+  public List<? extends UnboundedSource<FirebaseEvent<T>, FirebaseCheckpoint<T>>>
+    generateInitialSplits(
+      int numSplits, PipelineOptions arg1) throws Exception {
+    return Collections.singletonList(this);
+  }
+
+  @Override
+  public Coder<FirebaseEvent<T>> getDefaultOutputCoder() {
+    return FirebaseEventCoder.of(new TypeDescriptor<FirebaseEvent<T>>(){}, this.clazz);
+  }
+
+  @Override
+  public void validate() {
+
+    AuthData result;
+
+    try {
+      result = auther.authenticate(this.getRef());
+    } catch (FirebaseException|InterruptedException e) {
+      result = null;
+      e.printStackTrace();
```

Comment:
use logging instead of printStackTrace.
re-interrupt currentThread.

why check for non-null (with an empty failure message if the check fails) rather than simply converting the FireBasedException into a RuntimeException and throwing it?